py-package-info/main.py

```diff
@@ -30,71 +41,140 @@ def set_output(name, value):
+        print(f"::debug::Done fetching {package_info['name']}")
+    return checksum
+
+
+def lookup_package(name, check_test_index, version=None, attempts_limit=3):
+    package_metadata = None
+    package_info = {}
+    artifact = None
 
     package_index_url = "https://pypi.io/pypi/{}/json"
     if check_test_index:
         package_index_url = "https://test.pypi.org/pypi/{}/json"
 
     print(f"::debug::Checking the following package index {package_index_url}")
 
-    with closing(urlopen(package_index_url.format(name))) as f:
-        reader = codecs.getreader("utf-8")
-        pkg_data = json.load(reader(f))
-    if pkg_data is None:
-        raise RuntimeError(f"no package data for: {name}")
-    d = {}
-    d['name'] = pkg_data['info']['name']
-    d['version'] = pkg_data['info'].get('version', '')
-    d['homepage'] = pkg_data['info'].get('home_page', '')
-    d['summary'] = pkg_data['info'].get('summary', '')
-    d['author'] = pkg_data['info'].get('author', '')
-    d['author_email'] = pkg_data['info'].get('author_email', '')
-    artifact = None
-    if version:
-        for pypi_version in pkg_data['releases']:
-            if pkg_resources.safe_version(pypi_version) == version:
-                for version_artifact in pkg_data['releases'][pypi_version]:
-                    if version_artifact['packagetype'] == 'sdist':
-                        artifact = version_artifact
-                        d['version'] = version
-                        break
-        if artifact is None:
-            print("::warning::Could not find an exact version match for "
-                  f"{name} version {version}; using newest instead")
-
-    if artifact is None:  # no version given or exact match not found
-        for url in pkg_data['urls']:
-            if url['packagetype'] == 'sdist':
-                artifact = url
-                break
+    for attempt in range(attempts_limit):
+        print(
+            f"::debug::Fetching package metadata - attempt {attempt + 1} / {attempts_limit}")
+        try:
+            package_metadata = fetch_package_data(name, package_index_url)
+
+            if package_metadata is None:
+                raise PackageMetadataNotFoundInPyPIError(
+                    f"Could not find package metadata for: {name}")
+            package_info = get_package_info(package_metadata)
+
+            if version:
+                artifact = get_artifact_version(package_metadata, version)
+                if artifact is None:
+                    raise PackageVersionNotFoundInPyPIError("Could not find an exact version match for "
+                                                            f"{name} version {version}")
+                else:
+                    package_info['version'] = version
+        except Exception as e:
+            if attempt == attempts_limit - 1 and not isinstance(e, PackageVersionNotFoundInPyPIError):
+                raise RuntimeError(f"{e}")
+            elif attempt < attempts_limit - 1:
```

Comment:
Since we can't hit this if the attempt limit is past, drop the else.
```suggestion
            if attempt < attempts_limit - 1:
```